presto-hive/src/main/java/com/facebook/presto/hive/HiveS3Optimizer.java

```diff
@@ -0,0 +1,204 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.hive;
+
+import com.facebook.presto.hive.metastore.Column;
+import com.facebook.presto.hive.metastore.Partition;
+import com.facebook.presto.hive.metastore.Table;
+import com.facebook.presto.spi.ConnectorSession;
+import com.facebook.presto.spi.type.StandardTypes;
+import com.google.common.collect.ImmutableSet;
+import io.airlift.log.Logger;
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.io.compress.CompressionCodec;
+import org.apache.hadoop.io.compress.GzipCodec;
+import org.apache.hadoop.mapred.InputFormat;
+import org.apache.hadoop.mapred.TextInputFormat;
+
+import java.util.List;
+import java.util.Locale;
+import java.util.Optional;
+import java.util.Properties;
+import java.util.Set;
+
+import static com.facebook.presto.hive.metastore.MetastoreUtil.getHiveSchema;
+import static org.apache.hadoop.hive.metastore.api.hive_metastoreConstants.META_TABLE_LOCATION;
+
+/**
+ * HiveS3Optimizer uses Amazon S3 Select to push down queries to Amazon S3. This allows presto to retrieve only a
+ * subset of data rather than retrieving the full S3 object thus improving Presto query performance.
+ */
+public class HiveS3Optimizer
+{
+    private static final Logger log = Logger.get(HiveS3Optimizer.class);
+    private static final String S3_PREFIX = "s3://";
+
+    private static Set<String> supportedSerdes = ImmutableSet.of(
+            "org.apache.hadoop.hive.serde2.lazy.LazySimpleSerDe");
+
+    private static Set<String> supportedInputFormats = ImmutableSet.of(
+            "org.apache.hadoop.mapred.TextInputFormat");
+
+    private static Set<String> supportedColumnTypes = ImmutableSet.of(
+            "boolean",
+            "int",
+            "tinyint",
+            "smallint",
+            "bigint",
+            "string",
+            "float",
+            "double",
+            "decimal",
+            "date",
+            "timestamp");
```

Comment:
Do you support timestamps in legacy or non-legacy mode?

Reply:
We found some issues with the integration of timestamp data type. Thus, currently the optimizer doesn't push down timestamp predicates to S3Select. The exclusion logic is in QueryBuilder class. 

```
+    boolean shouldPushDown(Domain domain)
+    {
+        Type type = domain.getValues().getType();
+        /*
+         * Pushing down timestamp to s3select is problematic due to following reasons:
+         * 1) Presto bug: TIMESTAMP behaviour does not match sql standard (https://github.com/prestodb/presto/issues/7122)
+         * 2) Presto uses the timezone from client to convert the timestamp if no timezone is provided, however, s3select is a different service and this could lead to unexpected results.
+         * 3) ION SQL compare timestamps using precision, timestamps with different precisions are not equal even actually they present the same instant of time. This could lead to unexpected results.
+         */
+        return !(type instanceof TimestampType);
+    }
```

Reply:
Thanks